### ISSUE TITLE
No support for simd_mod<int64_t> in fflas_freduce?

### ISSUE THREAD

**Issue by P1K** [linbox-team/fflas-ffpack#224]
The current implementation of `fflas_freduce` does not seem to support vectorization for int64_t:

In `fflas_freduce.h`:
```
#ifdef SIMD_INT
    template<>
    struct support_simd_mod<int64_t> : public std::true_type {} ;
#endif  // SIMD_INT
```
At the bottom of `fflas_simd.h`:
```
#undef SIMD_INT
```
Trying to use simd for int64_t indeed immediately triggers compilation errors, so some work seems necessary to make this available.
I'm willing to investigate further, if you think that's worth it?

**Comment by cyrilbouvier** [comment]
Do you have an example that compiles of using `fflas_freduce` for integral type other than `int64_t` ?

I am currently trying to understand how the `simd::mod` method works and I found some usage of it in `fflas_freduce.inl` that seems weird in the integral case. For example, line 549
```cpp
C = SimdT::template mod<false, false, H.shift>( C, H.P, H.M,  H.NEGP, H.MIN, H.MAX, H.Q, H.T );
``` 
but the `mod` method of the Simd structs is only templated in `simd128_int64.inl`, `simd256_int64.inl` and `simd512_int64.inl`.

So it seems to me that `fflas_freduce` should not compile for any integral type other than `int64_t` but I was not able to produce a small piece of code using `fflas_freduce`.

**Comment by P1K** [comment]
A test that relies on `fflas_freduce` is `test_fgemm`.

I looked at this this morning with @Breush, and the conclusions were:
- the current code is flawed, because `H.shift` has to be known at compile-time, which is not the case here (`H.shift` is used as an argument to vectorized shift, whose shift operand is required to be an immediate)
- even by patching this (we just temporarily used an awful switch with 64 cases and 64 corresponding instantiations of mod with a constant switch), the result doesn't seem to be correct, so maybe the entire function is buggy.

I'm now planning to debug the mod function.

P.S. my WIP commits will be in the branch "looking_at_mulhi_fast", whose name isn't quite explicit. It also contains a preliminary vectorized implementation of `mulhi`, which might be better than what exists on AVX-512.

**Comment by P1K** [comment]
(Sorry, I don't think I answered your question. So no, I don't know of any case were `simd::mod` is used with something else than `int64_t`.
Actually, even this case was not activated in `fflas_freduce` (so no compilation error; it simply uses "scalar" mod), so I think that `simd::mod` is currently not used at all for integral types.)

**Comment by cyrilbouvier** [comment]
> I'm now planning to debug the mod function.

Ok I'll wait for your branch to be merged before trying to use the mod function with simd.

**Comment by P1K** [comment]
I have a seemingly working implementation of a vectorised mod for `int64_t` for both AVX and AVX-512 in the branch `simd_int64_mod`. This seems to improve performance quite a bit for largish inputs, as per the following benchmark using `benchmark-fgemm` on AVX-512 (results for AVX are kinda similar):
Original scalar:
```
Time: 0.247747 Gfops: 63.5223 -q 1048573 -m 1989 -k 1989 -n 1989 -w 0 -i 7 -p 0 -t 24 -b 24
Time: 0.251836 Gfops: 62.4908 -q 2097143 -m 1989 -k 1989 -n 1989 -w 0 -i 7 -p 0 -t 24 -b 24
Time: 0.256014 Gfops: 61.471 -q 4194301 -m 1989 -k 1989 -n 1989 -w 0 -i 7 -p 0 -t 24 -b 24
Time: 0.269454 Gfops: 58.405 -q 8388593 -m 1989 -k 1989 -n 1989 -w 0 -i 7 -p 0 -t 24 -b 24
Time: 0.379014 Gfops: 41.5221 -q 16777213 -m 1989 -k 1989 -n 1989 -w 0 -i 7 -p 0 -t 24 -b 24
Time: 0.3787 Gfops: 41.5565 -q 33554393 -m 1989 -k 1989 -n 1989 -w 0 -i 7 -p 0 -t 24 -b 24
Time: 0.378866 Gfops: 41.5383 -q 67108859 -m 1989 -k 1989 -n 1989 -w 0 -i 7 -p 0 -t 24 -b 24
Time: 0.378814 Gfops: 41.544 -q 134217689 -m 1989 -k 1989 -n 1989 -w 0 -i 7 -p 0 -t 24 -b 24
Time: 0.481613 Gfops: 32.6766 -q 268435399 -m 1989 -k 1989 -n 1989 -w 0 -i 7 -p 0 -t 24 -b 24
Time: 0.885487 Gfops: 17.7726 -q 536870909 -m 1989 -k 1989 -n 1989 -w 0 -i 7 -p 0 -t 24 -b 24
Time: 2.43931 Gfops: 6.4516 -q 1073741789 -m 1989 -k 1989 -n 1989 -w 0 -i 7 -p 0 -t 24 -b 24
Time: 8.83502 Gfops: 1.78126 -q 2147483647 -m 1989 -k 1989 -n 1989 -w 0 -i 7 -p 0 -t 24 -b 24
```

Vectorised:
```
Time: 0.247179 Gfops: 63.6682 -q 1048573 -m 1989 -k 1989 -n 1989 -w 0 -i 7 -p 0 -t 24 -b 24
Time: 0.253374 Gfops: 62.1115 -q 2097143 -m 1989 -k 1989 -n 1989 -w 0 -i 7 -p 0 -t 24 -b 24
Time: 0.247416 Gfops: 63.6073 -q 4194301 -m 1989 -k 1989 -n 1989 -w 0 -i 7 -p 0 -t 24 -b 24
Time: 0.26823 Gfops: 58.6715 -q 8388593 -m 1989 -k 1989 -n 1989 -w 0 -i 7 -p 0 -t 24 -b 24
Time: 0.347072 Gfops: 45.3435 -q 16777213 -m 1989 -k 1989 -n 1989 -w 0 -i 7 -p 0 -t 24 -b 24
Time: 0.347958 Gfops: 45.228 -q 33554393 -m 1989 -k 1989 -n 1989 -w 0 -i 7 -p 0 -t 24 -b 24
Time: 0.348473 Gfops: 45.1612 -q 67108859 -m 1989 -k 1989 -n 1989 -w 0 -i 7 -p 0 -t 24 -b 24
Time: 0.347443 Gfops: 45.295 -q 134217689 -m 1989 -k 1989 -n 1989 -w 0 -i 7 -p 0 -t 24 -b 24
Time: 0.370812 Gfops: 42.4405 -q 268435399 -m 1989 -k 1989 -n 1989 -w 0 -i 7 -p 0 -t 24 -b 24
Time: 0.429817 Gfops: 36.6143 -q 536870909 -m 1989 -k 1989 -n 1989 -w 0 -i 7 -p 0 -t 24 -b 24
Time: 0.626793 Gfops: 25.1079 -q 1073741789 -m 1989 -k 1989 -n 1989 -w 0 -i 7 -p 0 -t 24 -b 24
Time: 1.63201 Gfops: 9.64298 -q 2147483647 -m 1989 -k 1989 -n 1989 -w 0 -i 7 -p 0 -t 24 -b 24
```

**Comment by P1K** [comment]
(The main zone of interest is probably the last four primes, for which modular balance double can't be used.)

**Comment by ClementPernet** [comment]
This is awesome! Depending on what you mean by 'seemingly', I suggest to PR this branch!

**Comment by P1K** [comment]
Means that it passes all the tests I've done so far, but I haven't done too many yet ^^
I think it should be mergeable quite soon. Beyond a bit more testing, I need to do Simd128, and we need to decide what to do with the previous code that wasn't used and didn't seem to work (keep it as is/comment it/remove it?).
There's also the point raised by Cyril above: are we fine with implementing integral`simd::mod` only for `int64_t`, or do we want others (e.g. to allow `fflas_freduce` to use different integral types)?

**Comment by P1K** [comment]
(Some more benchs, with pluq on AVX2:
- `ModularBalanced<double>`
```
Time: 0.403357 Gfops: 13.2224 -s N -q 8388593 -m 2000 -n 2000 -r 2000 -g Y -i 7 -v 0 -t 1 -b 1 -p N
Time: 0.423803 Gfops: 12.5845 -s N -q 16777213 -m 2000 -n 2000 -r 2000 -g Y -i 7 -v 0 -t 1 -b 1 -p N
Time: 0.47652 Gfops: 11.1923 -s N -q 33554393 -m 2000 -n 2000 -r 2000 -g Y -i 7 -v 0 -t 1 -b 1 -p N
Time: 0.710614 Gfops: 7.50525 -s N -q 67108859 -m 2000 -n 2000 -r 2000 -g Y -i 7 -v 0 -t 1 -b 1 -p N
Time: 1.65921 Gfops: 3.21438 -s N -q 134217689 -m 2000 -n 2000 -r 2000 -g Y -i 7 -v 0 -t 1 -b 1 -p N
```
- `ModularBalanced<int64_t>` scalar mod
```
Time: 0.34393 Gfops: 15.507 -s N -q 8388593 -m 2000 -n 2000 -r 2000 -g Y -i 7 -v 0 -t 1 -b 1 -p N
Time: 0.651034 Gfops: 8.1921 -s N -q 16777213 -m 2000 -n 2000 -r 2000 -g Y -i 7 -v 0 -t 1 -b 1 -p N
Time: 0.533071 Gfops: 10.0049 -s N -q 33554393 -m 2000 -n 2000 -r 2000 -g Y -i 7 -v 0 -t 1 -b 1 -p N
Time: 0.543154 Gfops: 9.81919 -s N -q 67108859 -m 2000 -n 2000 -r 2000 -g Y -i 7 -v 0 -t 1 -b 1 -p N
Time: 0.55545 Gfops: 9.60182 -s N -q 134217689 -m 2000 -n 2000 -r 2000 -g Y -i 7 -v 0 -t 1 -b 1 -p N
Time: 0.598977 Gfops: 8.90407 -s N -q 268435399 -m 2000 -n 2000 -r 2000 -g Y -i 7 -v 0 -t 1 -b 1 -p N
Time: 0.768012 Gfops: 6.94434 -s N -q 536870909 -m 2000 -n 2000 -r 2000 -g Y -i 7 -v 0 -t 1 -b 1 -p N
Time: 1.62153 Gfops: 3.28907 -s N -q 1073741789 -m 2000 -n 2000 -r 2000 -g Y -i 7 -v 0 -t 1 -b 1 -p N
Time: 5.31568 Gfops: 1.00332 -s N -q 2147483647 -m 2000 -n 2000 -r 2000 -g Y -i 7 -v 0 -t 1 -b 1 -p N
```
- `ModularBalanced<int64_t>` vectorised mod
```
Time: 0.351221 Gfops: 15.1851 -s N -q 8388593 -m 2000 -n 2000 -r 2000 -g Y -i 7 -v 0 -t 1 -b 1 -p N
Time: 0.343722 Gfops: 15.5164 -s N -q 16777213 -m 2000 -n 2000 -r 2000 -g Y -i 7 -v 0 -t 1 -b 1 -p N
Time: 0.330589 Gfops: 16.1328 -s N -q 33554393 -m 2000 -n 2000 -r 2000 -g Y -i 7 -v 0 -t 1 -b 1 -p N
Time: 0.42375 Gfops: 12.586 -s N -q 67108859 -m 2000 -n 2000 -r 2000 -g Y -i 7 -v 0 -t 1 -b 1 -p N
Time: 0.401785 Gfops: 13.2741 -s N -q 134217689 -m 2000 -n 2000 -r 2000 -g Y -i 7 -v 0 -t 1 -b 1 -p N
Time: 0.36046 Gfops: 14.7959 -s N -q 268435399 -m 2000 -n 2000 -r 2000 -g Y -i 7 -v 0 -t 1 -b 1 -p N
Time: 0.396188 Gfops: 13.4616 -s N -q 536870909 -m 2000 -n 2000 -r 2000 -g Y -i 7 -v 0 -t 1 -b 1 -p N
Time: 0.599749 Gfops: 8.89261 -s N -q 1073741789 -m 2000 -n 2000 -r 2000 -g Y -i 7 -v 0 -t 1 -b 1 -p N
Time: 1.51539 Gfops: 3.51946 -s N -q 2147483647 -m 2000 -n 2000 -r 2000 -g Y -i 7 -v 0 -t 1 -b 1 -p N
```
)

**Comment by P1K** [comment]
Closed by merge #234
